### PR TITLE
PLAT-738 Add database structure and migration resources

### PR DESCRIPTION
**Pull request by alexander-schmidt**
- Added resource files for JSON-based database structure versions, and corresponding SQL-based migration scripts.
  - Created sub-project `platform-integration` which depends on all other platform projects.
  - Created `DatabaseStructureVersionCodeValidator` to enforce that the most recent JSON based database structure file corresponds to the platform tables in the classpath.
  - Added resource files for platform-15 up to platform-17.
- `JavaCodeValidationEnvironment` can now handle JSON based configuration parameters.
  - Added `jsonPath` dependency. This entailed upgrades of `asm` and `slf4j`.
  - Created `JsonValueReader[Test]`.
- Unrelated:
  - Improved imprecise regexes in `ClasspathResourcesValidator`.

### CODE REVIEW

File: platform-common/src/main/java/com/softicar/platform/common/io/resource/supplier/IResourceSupplier.java [hunk 1]
```diff
@@ -28,6 +32,22 @@
...
 	IResourceKey getResourceKey();
 
+	/**
+	 * Returns the content of the {@link IResource} as UTF-8 encoded text.
+	 * <p>
+	 * May only be called on textual resources.
+	 *
+	 * @return the content as UTF-8 encoded text (never <i>null</i>)
+	 */
+	default String getTextUtf8() {
```
Comment (oliver-richers) [inline platform-common/src/main/java/com/softicar/platform/common/io/resource/supplier/IResourceSupplier.java:42]: uh, I think this should rather be implemented on IResource, not the supplier

File: platform-common/src/main/java/com/softicar/platform/common/io/serialization/json/JsonValueReader.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	 * {@link String}.
+	 * <p>
+	 * Returns an empty {@link Optional} if no such path exists inside the JSON
+	 * {@link String}, or if the path points to an array of values.
+	 *
+	 * @param path
+	 *            the path, in "JsonPath" syntax (never <i>null</i>)
+	 * @return the {@link String} value to which the given path points
+	 */
+	public Optional<String> read(String path) {
```
Comment (oliver-richers) [inline platform-common/src/main/java/com/softicar/platform/common/io/serialization/json/JsonValueReader.java:48]: maybe `readValue ` would be a better name to constrast it to the `readList` method?

File: platform-common/src/test/java/com/softicar/platform/common/io/serialization/json/JsonValueReaderTest.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+	@Test
+	public void testReadListWithValueArray() {
+
+		var reader = createReader();
+		List<String> valueList = reader.readList("$.outerElement.innerArray");
+		assertFalse(valueList.isEmpty());
+		assertEquals("first", valueList.get(0));
+		assertEquals("second", valueList.get(1));
+		assertEquals("third", valueList.get(2));
```
Comment (oliver-richers) [inline platform-common/src/test/java/com/softicar/platform/common/io/serialization/json/JsonValueReaderTest.java:63]: tip: asserts could be simplified to `assertEquals("[first, second, third]", valueList.toString());`

File: platform-integration/.settings/com.softicar.sqml.Sqml.prefs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+BuilderConfiguration.is_project_specific=true
```
Comment (oliver-richers) [inline platform-integration/.settings/com.softicar.sqml.Sqml.prefs:1]: hmm, should this project ever contain SQML? maybe remove the nature?

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/DatabaseStructureJsonFromClasspathExtractor.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	private String determineCreateStatement(IDbTable<?, ?> table) {
+
+		var structure = new DbRuntimeTableStructure(table);
+		var sqlGenerator = new DbTableStructureSqlGenerator(structure);
+		return sqlGenerator.getCreateTableStatement();
+	}
+
+	private Collection<IDbTable<?, ?>> loadUniqueTables() {
```
Comment (oliver-richers) [inline platform-integration/src/main/java/com/softicar/platform/integration/database/structure/DatabaseStructureJsonFromClasspathExtractor.java:53]: how do we ensure that the integration project depends on **all** other project, to find **all** tables?

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResource.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+ * extensions).
+ * <p>
+ * By definition, {@code "vX-migration.sql"} contains the necessary statements
+ * to migrate the structure and content of an existing database instance from
+ * version {@code "X-1"} to version {@code "X"}.
+ *
+ * @author Alexander Schmidt
+ */
+@ResourceSupplierContainer
+public interface DatabaseStructureVersionResource {
```
Comment (oliver-richers) [inline platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResource.java:24]: why is the name of this class singular? I would expect plural

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import com.softicar.platform.common.core.exceptions.SofticarDeveloperException;
+import com.softicar.platform.common.io.resource.supplier.IResourceSupplier;
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class DatabaseStructureVersionResources {
```
Comment (oliver-richers) [inline platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java:12]: isn't this unnecessarily complex? why use reflection? wouldn't a simple List of Strings suffice? or a text file listing all resources?

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-common/src/main/java/com/softicar/platform/common/io/resource/supplier/IResourceSupplier.java (hunk 1 above)
Comment (alexander-schmidt) [reply platform-common/src/main/java/com/softicar/platform/common/io/resource/supplier/IResourceSupplier.java:42]: indeed

File: platform-common/src/main/java/com/softicar/platform/common/io/serialization/json/JsonValueReader.java (hunk 2 above)
Comment (alexander-schmidt) [reply platform-common/src/main/java/com/softicar/platform/common/io/serialization/json/JsonValueReader.java:48]: ok

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/DatabaseStructureJsonFromClasspathExtractor.java (hunk 5 above)
Comment (alexander-schmidt) [reply platform-integration/src/main/java/com/softicar/platform/integration/database/structure/DatabaseStructureJsonFromClasspathExtractor.java:53]: So far, we don't. From the top of my mind, could determine a list of all sub-projects in main `build.gradle` script, and somehow compare that to the project configuration of the `platform-integration` sub-project. Can you see an elegant way to do this (or something equivalent)?

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java (hunk 7 above)
Comment (alexander-schmidt) [reply platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java:12]: I can't see how to implement `getLatestStructureResourceSupplier` or `getAllStructureResourceSuppliers` that way. Do you have a specific approach in mind? If so, please provide a (pseudo-)code snippet.

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResource.java (hunk 6 above)
Comment (alexander-schmidt) [reply platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResource.java:24]: The pluralized name is occupied by the corresponding utility class.
Btw, a search for `ResourceSupplierContainer` references revealed that we currently use singular and plural alike, for these types. So this is not an isolated case.

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/DatabaseStructureJsonFromClasspathExtractor.java (hunk 5 above)
Comment (oliver-richers) [reply platform-integration/src/main/java/com/softicar/platform/integration/database/structure/DatabaseStructureJsonFromClasspathExtractor.java:53]: nope, but we might run into problems, just saying

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java (hunk 7 above)
Comment (oliver-richers) [reply platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java:12]: well some file like `database-migrations` with content
```
v15-structure.json
v16-migration.sql
...
```

Or instead some field in `getLatestStructureResourceSupplier` like
```
private static final String RESOURCES = """
v15-structure.json
v16-migration.sql
...
"""
```

just asking, we can leave it as is, but why more complicated than necessary?

**Review by oliver-richers — APPROVED** [review]
generally accepted, but check my comment

File: platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java (hunk 7 above)
Comment (oliver-richers) [reply platform-integration/src/main/java/com/softicar/platform/integration/database/structure/version/DatabaseStructureVersionResources.java:12]: ok, I forgot about that, all fine ;)

**Review by oliver-richers — APPROVED** [review]

**Pull request merged**